### PR TITLE
转换规则 torch.nn.optim.lr_scheduler

### PR DESCRIPTION
**Pull request by RedContritio**
### PR Docs

https://github.com/PaddlePaddle/docs/pull/5979

### PR APIs

```bash
torch.optim.lr_scheduler.CosineAnnealingLR
torch.optim.lr.scheduler.CyclicLR
torch.optim.lr_scheduler.ExponentialLR
torch.optim.lr_scheduler.LambdaLR
torch.optim.lr_scheduler.MultiStepLR
torch.optim.lr_scheduler.MultiplicativeLR
torch.optim.lr_scheduler.OneCycleLR
torch.optim.lr_scheduler.ReduceLROnPlateau
```

### CODE REVIEW

File: paconvert/api_matcher.py [hunk 1]
```diff
@@ -3570,29 +3578,51 @@ def generate_code(self, kwargs):
...
+            total_iters = kwargs.pop("total_iters")
+        kwargs["values"] = "[{}*{}.get_lr(), {}.get_lr()]".format(factor, optim, optim)
+        kwargs["boundaries"] = "[{}]".format(total_iters)
+        return super().generate_code(kwargs)
+
+
+class OneCycleLRMatcher(LRSchedulerMatcher):
+    def generate_code(self, kwargs):
+        if "total_steps" in kwargs and "None" not in kwargs["total_steps"]:
+            pass
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:3611]: 这个是支持转写还是不支持转写，尽量在json里配置，如果是无用的参数，可以使用 key: "" 的形式来配置·

File: tests/test_optim_lr_scheduler_ConstantLR.py [hunk 2]
```diff
@@ -52,13 +52,13 @@ def test_case_1():
         conv.bias = torch.nn.Parameter(bias)
 
         sgd = torch.optim.SGD(conv.parameters(), lr=0.1)
-        schduler = torch.optim.lr_scheduler.ConstantLR(sgd, verbose=True)
+        scheduler = torch.optim.lr_scheduler.ConstantLR(sgd, verbose=True)
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_ConstantLR.py:55]: 这个单测也可以用 `generate_torch_code` 的形式优化下，增加代码的复用性，然后增加几个case，目前2个偏少了

File: tests/test_optim_lr_scheduler_StepLR.py [hunk 3]
```diff
@@ -52,13 +52,13 @@ def test_case_1():
         conv.bias = torch.nn.Parameter(bias)
 
         sgd = torch.optim.SGD(conv.parameters(), lr=0.1)
-        schduler = torch.optim.lr_scheduler.StepLR(sgd, step_size=2, verbose=True)
+        scheduler = torch.optim.lr_scheduler.StepLR(sgd, step_size=2, verbose=True)
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_StepLR.py:55]: 同上，可以优化下代码复用性，然后增加点case，2个太偏少了

File: paconvert/api_matcher.py (hunk 1 above)
Comment (RedContritio) [reply paconvert/api_matcher.py:3611]: 这里的逻辑是，有三个参数 `total_steps`, `steps_per_epoch` 和 `epochs`。

规则是，如果设置了 `total_steps` （默认参数为 `None`），则直接使用 `total_steps` 传递给 paddle 代码；
如果未设置 `total_steps` （即不存在 `total_steps` 参数或设置为 `None`），则基于 `steps_per_epoch` 和 `epochs` 相乘作为 `total_steps` 传递给 paddle。

因为这个逻辑比较复杂，所以在这里单独处理了。

File: tests/test_optim_lr_scheduler_ConstantLR.py (hunk 2 above)
Comment (RedContritio) [reply tests/test_optim_lr_scheduler_ConstantLR.py]: 已解决

File: tests/test_optim_lr_scheduler_CyclicLR.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        )
+    )
+    obj.run(pytorch_code, ["result1", "result2"], rtol=1.0e-5, check_value=False)
+
+
+def test_case_6():
+    pytorch_code = textwrap.dedent(
+        generate_torch_code(
+            "torch.optim.lr_scheduler.CyclicLR(scale_fn=None, scale_mode='cycle', last_epoch=-1, verbose=False, base_lr=0.01, max_lr=0.1, step_size_up=1000, optimizer=sgd, mode='triangular', gamma=1.0)"
+        )
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_CyclicLR.py:121]: 可以不指定关键字测一个

File: tests/test_optim_lr_scheduler_ExponentialLR.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            "torch.optim.lr_scheduler.ExponentialLR(optimizer=sgd, gamma=0.5, last_epoch=-1, verbose=True)"
+        )
+    )
+    obj.run(pytorch_code, ["result1", "result2"], rtol=1.0e-5)
+
+
+def test_case_5():
+    pytorch_code = textwrap.dedent(
+        generate_torch_code(
+            "torch.optim.lr_scheduler.ExponentialLR(optimizer=sgd, gamma=0.5, verbose=True)"
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_ExponentialLR.py:103]: 不指定关键字测一个

File: tests/test_optim_lr_scheduler_LambdaLR.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+def test_case_3():
+    pytorch_code = textwrap.dedent(
+        generate_torch_code(
+            "torch.optim.lr_scheduler.LambdaLR(optimizer=sgd, lr_lambda=lambda x:0.95**x)"
+        )
+    )
+    obj.run(pytorch_code, ["result1", "result2"], rtol=1.0e-5)
+
+
+def test_case_4():
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_LambdaLR.py:93]: 全部不指定关键字测一个

File: tests/test_optim_lr_scheduler_MultiStepLR.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            "torch.optim.lr_scheduler.MultiStepLR(optimizer=sgd, milestones=[2,4,6])"
+        )
+    )
+    obj.run(pytorch_code, ["result1", "result2"], rtol=1.0e-5)
+
+
+def test_case_4():
+    pytorch_code = textwrap.dedent(
+        generate_torch_code(
+            "torch.optim.lr_scheduler.MultiStepLR(optimizer=sgd, milestones=[2,4,6], gamma=0.5, last_epoch=-1, verbose=True)"
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_MultiStepLR.py:96]: 全部不指定关键字

File: tests/test_optim_lr_scheduler_MultiplicativeLR.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+def test_case_3():
+    pytorch_code = textwrap.dedent(
+        generate_torch_code(
+            "torch.optim.lr_scheduler.MultiplicativeLR(optimizer=sgd, lr_lambda=lambda x:0.95**x)"
+        )
+    )
+    obj.run(pytorch_code, ["result1", "result2"], rtol=1.0e-5)
+
+
+def test_case_4():
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_MultiplicativeLR.py:95]: 全部不指定关键字

File: tests/test_optim_lr_scheduler_OneCycleLR.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+def test_case_5():
+    pytorch_code = textwrap.dedent(
+        generate_torch_code(
+            "torch.optim.lr_scheduler.OneCycleLR(optimizer=sgd, max_lr=0.01, steps_per_epoch=20, epochs=10, verbose=True)"
+        )
+    )
+    obj.run(pytorch_code, ["result1", "result2"], rtol=1.0e-5, check_value=False)
+
+
+def test_case_6():
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_OneCycleLR.py:111]: 全部不指定关键字都测一个

**Review by zhwesky2010 — COMMENTED** [review]
可以全部不要关键字测一个，至少包含以下用例：
```
1. 全部指定关键字
2. 全部不指定关键字
3. 默认参数都不输入
```

File: tests/test_optim_lr_scheduler_ConstantLR.py [hunk 10]
```diff
@@ -19,142 +19,93 @@
...
+def test_case_5():
+    pytorch_code = textwrap.dedent(
+        generate_torch_code(
+            "torch.optim.lr_scheduler.ConstantLR(optimizer=sgd, factor=0.05, total_iters=3)"
+        )
+    )
+    obj.run(pytorch_code, ["result1", "result2"], rtol=1.0e-5)
+
+
+def test_case_6():
```
[the remark this answers is not in the document]
Comment (RedContritio) [reply tests/test_optim_lr_scheduler_ConstantLR.py:105]: 这部分是因为主观判断 `last_epoch` 和 `verbose` 这两个参数意义不大，所以没有单独测试

File: tests/test_optim_lr_scheduler_CyclicLR.py (hunk 4 above)
Comment (RedContritio) [reply tests/test_optim_lr_scheduler_CyclicLR.py:121]: 这个里面 `cycle_momentum`, `base_momentum`, `max_momentum` 参数不支持，所以没法测试全部参数，只能加一个 “这三个参数之前的不指定关键字，跳过这三个，这三个之后的指定关键字”

File: tests/test_optim_lr_scheduler_OneCycleLR.py (hunk 9 above)
Comment (RedContritio) [reply tests/test_optim_lr_scheduler_OneCycleLR.py:111]: 这个同上，只能不指定大多数关键字

File: tests/test_optim_lr_scheduler_ConstantLR.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (RedContritio) [reply tests/test_optim_lr_scheduler_ConstantLR.py:105]: 已经加上

[the remark this answers is not in the document]
Comment (zhwesky2010) [reply tests/test_optim_lr_scheduler_ConstantLR.py:105]: > 这部分是因为主观判断 `last_epoch` 和 `verbose` 这两个参数意义不大，所以没有单独测试

以防万一还是测一下吧

File: tests/test_optim_lr_scheduler_OneCycleLR.py (hunk 9 above)
Comment (zhwesky2010) [reply tests/test_optim_lr_scheduler_OneCycleLR.py:111]: > 这个同上，只能不指定大多数关键字

OK

File: tests/test_optim_lr_scheduler_ConstantLR.py [hunk 11]
```diff
@@ -109,3 +109,21 @@ def test_case_6():
...
+def test_case_7():
+    pytorch_code = textwrap.dedent(
+        generate_torch_code(
+            "torch.optim.lr_scheduler.ConstantLR(sgd, 0.05, 3, -1, False)"
+        )
+    )
+    obj.run(pytorch_code, ["result1", "result2"], rtol=1.0e-5)
+
+
+def test_case_8():
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_ConstantLR.py:123]: 其他没问题，就是last_epoch这个参数，看目前测的last_epoch都为-1，如果不为-1，这个参数对应一致不

Comment (RedContritio) [reply tests/test_optim_lr_scheduler_ConstantLR.py:78]: `torch.optim.lr_scheduler.ReduceLROnPlateau` 函数签名不含 `last_epoch` 参数，跳过；
部分 scheduler 使用 `last_epoch` 得到的结果值不相同，已使用 `check_value=False` 忽略。

File: tests/test_optim_lr_scheduler_ReduceLROnPlateau.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        generate_torch_code(
+            "torch.optim.lr_scheduler.ReduceLROnPlateau(sgd, 'max', 0.1, 10, 1e-4, 'rel', 0, 0, 1e-8, False)",
+            step_with_loss=True,
+        )
+    )
+    obj.run(pytorch_code, ["result1", "result2"], rtol=1.0e-5)
+
+
+# note: fine, ReduceLROnPlateau does not support `last_epoch`
+def test_case_8():
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_ReduceLROnPlateau.py:93]: 这个后面可以直接删掉

File: tests/test_optim_lr_scheduler_StepLR.py [hunk 13]
```diff
@@ -15,146 +15,88 @@
...
+        generate_torch_code(
+            "torch.optim.lr_scheduler.StepLR(step_size=3, gamma=0.2, optimizer=sgd)"
+        )
+    )
+    obj.run(pytorch_code, ["result1", "result2"], rtol=1.0e-5)
+
+
+# note: StepLR does not support resume training
+# paddle result has diff with pytorch result
+def test_case_9():
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_StepLR.py:93]: 直接设置last_epoch=10来计算吧，排除下last_epoch更新不一致的其他原因

File: tests/test_optim_lr_scheduler_CyclicLR.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+from lr_scheduler_helper import generate_torch_code
+
+obj = APIBase("torch.optim.lr_scheduler.CyclicLR")
+
+
+def test_case_1():
+    pytorch_code = textwrap.dedent(
+        generate_torch_code("torch.optim.lr_scheduler.CyclicLR(sgd, 0.01, 0.1)")
+    )
+    obj.run(pytorch_code, ["result1", "result2"], rtol=1.0e-5, check_value=False)
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_CyclicLR.py:27]: 这个API是所有结果都对不齐吗

File: tests/test_optim_lr_scheduler_OneCycleLR.py [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+from lr_scheduler_helper import generate_torch_code
+
+obj = APIBase("torch.optim.lr_scheduler.OneCycleLR")
+
+
+def test_case_1():
+    pytorch_code = textwrap.dedent(
+        generate_torch_code("torch.optim.lr_scheduler.OneCycleLR(sgd, 0.01, 100)")
+    )
+    obj.run(pytorch_code, ["result1", "result2"], rtol=1.0e-5, check_value=False)
```
Comment (zhwesky2010) [inline tests/test_optim_lr_scheduler_OneCycleLR.py:27]: 这个API是所有结果都对不齐吗

File: tests/test_optim_lr_scheduler_StepLR.py (hunk 13 above)
Comment (RedContritio) [reply tests/test_optim_lr_scheduler_StepLR.py:93]: 这部分是直接用的上一次的 `scheduler` 作为 `last_epoch` 的，两边都能正常运行的话应该是正常获取的

File: tests/test_optim_lr_scheduler_CyclicLR.py (hunk 14 above)
Comment (RedContritio) [reply tests/test_optim_lr_scheduler_CyclicLR.py:27]: 是的，因为有随机因素

File: tests/test_optim_lr_scheduler_OneCycleLR.py (hunk 15 above)
Comment (RedContritio) [reply tests/test_optim_lr_scheduler_OneCycleLR.py:27]: 是的，这个也有随机因素

File: tests/test_optim_lr_scheduler_ReduceLROnPlateau.py (hunk 12 above)
Comment (RedContritio) [reply tests/test_optim_lr_scheduler_ReduceLROnPlateau.py:93]: 好的了解

**Review by zhwesky2010 — APPROVED** [review]
LGTM
